File: ios/Classes/TiReviewdialogModule.m
/**
 * ti.reviewdialog
 *
 * Created by Hans Knoechel
 * Copyright (c) 2017-present Hans Knoechel. All rights reserved.
 */

#import "TiReviewdialogModule.h"
#import <StoreKit/StoreKit.h>

@implementation TiReviewdialogModule

#pragma mark Internal

-(id)moduleGUID
{
	return @"6f1b3b5d-db05-47da-9e7d-ebb4124cdd8d";
}

-(NSString*)moduleId
{
	return @"ti.reviewdialog";
}

#pragma Public APIs

- (id)isSupported:(id)unused
{
    return @(YES); // The module's minimum target is iOS 11 now
}

- (void)requestReview:(id)unused
{
    ENSURE_UI_THREAD(requestReview, unused);    
    [SKStoreReviewController requestReview];
}

@end

File: ios/Classes/TiReviewdialogModule.swift
//
//  TiReviewdialogModule.swift
//  titanium-review-dialog
//
//  Created by Hans Knoechel
//  Copyright (c) 2017-present. All rights reserved.
//

import UIKit
import TitaniumKit
import StoreKit

@objc(TiReviewdialogModule)
class TiReviewdialogModule: TiModule {
  
  public let testProperty: String = "Hello World"
  
  func moduleGUID() -> String {
    return "977d9a13-62a1-4d3a-b907-131240f744e6"
  }
  
  override func moduleId() -> String! {
    return "ti.reviewdialog"
  }
  
  @objc(isSupported:)
  func isSupported(unused: [Any]?) -> Bool {
    return true
  }
  
  @objc(requestReview:)
  func requestReview(unused: [Any]?) {
    let window = UIApplication.shared.connectedScenes.compactMap { ($0 as? UIWindowScene)?.keyWindow }.last
    guard let scene = window?.windowScene else {
      return
    }
    
    if #available(iOS 16.0, *) {
      Task {
        await AppStore.requestReview(in: scene)
      }
    } else {
      SKStoreReviewController.requestReview(in: scene)
    }
  }
}
